lib/samly/state/redis.ex

```diff
@@ -0,0 +1,164 @@
+  It supports both direct Redix connections and custom Redis client modules.
+
+  ## Options
+
+  +   `:redis_module` - Redis client module (optional)
+                        Value must be an atom (module name)
+                        Defaults to `Redix`
+                        When using a custom module, it must implement `command/1` or `command/2`
+  +   `:redis_name` - Redix connection name (optional, only used with Redix)
+                      Value must be an atom
+                      Defaults to `:redix`
+  +   `:key_prefix` - Prefix for Redis keys (optional)
+                      Value must be a string
+                      Defaults to `"samly:assertion:"`
+  +   `:ttl` - Time to live in seconds for assertions (optional)
+               Value must be a positive integer
+               Defaults to `3600`
+
+  ## Configuration Examples
+
+  ### Using Redix directly (default)
+
+      config :samly, Samly.State,
+        opts: [
+          redis_name: :my_redix,
+          key_prefix: "samly:assertion:",
+          ttl: 60
+        ]
+
+  ### Using a custom Redis module (e.g., App.Redis with connection pool)
+
+      config :samly, Samly.State,
+        opts: [
+          redis_module: App.Redis,
+          key_prefix: "samly:assertion:",
+          ttl: 60
+        ]
+  """
+
+  alias Samly.Assertion
+
+  @behaviour Samly.State.Store
+
+  @default_redis_module Redix
+  @default_redis_name :redix
+  @default_key_prefix "samly:assertion:"
+  @default_ttl 3600
+
+  @impl Samly.State.Store
+  def init(opts) do
+    redis_module = Keyword.get(opts, :redis_module, @default_redis_module)
+    redis_name = Keyword.get(opts, :redis_name, @default_redis_name)
+    key_prefix = Keyword.get(opts, :key_prefix, @default_key_prefix)
+    ttl = Keyword.get(opts, :ttl, @default_ttl)
+
+    if not is_atom(redis_module) do
+      raise "Samly.State.Redis redis_module must be an atom: #{inspect(redis_module)}"
+    end
+
+    if not is_atom(redis_name) do
+      raise "Samly.State.Redis redis_name must be an atom: #{inspect(redis_name)}"
+    end
+
+    if not is_binary(key_prefix) do
+      raise "Samly.State.Redis key_prefix must be a string: #{inspect(key_prefix)}"
+    end
+
+    if ttl != nil and (not is_integer(ttl) or ttl <= 0) do
+      raise "Samly.State.Redis ttl must be a positive integer or nil: #{inspect(ttl)}"
+    end
+
+    [
+      redis_module: redis_module,
+      redis_name: redis_name,
+      key_prefix: key_prefix,
+      ttl: ttl
+    ]
+  end
+
+  @impl Samly.State.Store
+  def get_assertion(_conn, nil, _opts), do: nil
+
+  def get_assertion(_conn, assertion_key, opts) do
+    key_prefix = Keyword.fetch!(opts, :key_prefix)
+    redis_key = build_redis_key(key_prefix, assertion_key)
+
+    case redis_command(opts, ["GET", redis_key]) do
+      {:ok, nil} ->
+        nil
+
+      {:ok, binary} when is_binary(binary) ->
+        binary
+        |> Base.decode64!()
+        |> :erlang.binary_to_term()
```

Comment:
nit: i'd probably good to use the `:safe` option when loading the binary, wdyt?

https://www.erlang.org/doc/apps/erts/erlang.html#binary_to_term/2

> The safe option prevents direct and indirect creation of new atoms (such as those embedded in certain structures like process identifiers) and creation of new external function references. None of these resources are garbage collected, so unchecked creation can exhaust available memory.